train.py

```diff
@@ -81,6 +72,14 @@ def main(job_config: JobConfig):
     if parallel_dims.pp_enabled:
         pp_mesh = world_mesh["pp"]
 
+    # Set random seed, and maybe enable deterministic mode (mainly for debugging, expect perf loss)
+    utils.set_determinism(
+        spmd_mesh=world_mesh["dp", "tp"],
```

Comment:
note that this only works for 3D (PP+DP+TP). Either "dp" or "tp" being absent causes error.

Reply:
I think this assumption is right. Though I'm not sure if this still holds for the EP coming. It can be directly added in `train.py` because this is likely the only place this logic gets used. Besides, `DeviceMesh` should not be aware of parallelisms IMO.

Reply:
Yes, EP should still be SPMD world and the design of `DeviceMesh` is that it can be used without parallelisms. So it is better to have a util outside of `DeviceMesh`.

Reply:
yea, agree that device mesh shouldn't be aware of parellelisms.  Just wondered if device mesh should provide a helper API for filtering dimensions. Maybe not worth adding.  This is the code I wrote in train.py- and I wish it were a little cleaner:

`spmd_mesh_dims = list(filter(lambda name: name != "pp", world_mesh.mesh_dim_names))`